### DISCUSSION TITLE
Retries for specific gRPC errors/error codes

### DISCUSSION THREAD

**Discussion by vaibhavsawhney** [linkerd/linkerd2#4378] [General] [answered]
As I can see that we used to have [`gRPC Response Classifiers`](https://api.linkerd.io/head/linkerd/#grpc-response-classifiers) in Linkerd 1.x through which we could decide which type of errors/ error codes we want to retry our request for if service profile is enabled. 

I tried finding out if we can do something similar to this in Linkerd 2.x but I couldn't. 
The main motive of using this was to not retry for errors which are irrecoverable in nature but Linkerd retries for them as well.

I could find [`Response Match`](https://linkerd.io/2/reference/service-profiles/#response-match) which can be used to define status codes for considering a response as failure and success. But then I looked into an issue https://github.com/linkerd/linkerd2/issues/3376 raised last year around the same for which a fix was done later https://github.com/linkerd/linkerd2/issues/3729. Though this fix has a fixed set of error codes that will be retried and cannot be changed according to the client's preference.

I wanted to ask if there is any further development in this area. Is there a way to define error codes that should be retried?

**Comment by grampelberg** [comment, marked as answer]
@adleong we can do HTTP error codes in service profiles but not gRPC error codes right?

**Reply by vaibhavsawhney** [reply]
Yes that's what the issue https://github.com/linkerd/linkerd2/issues/3376 says and I too couldn't get any other way for gRPC.

**Reply by olix0r** [reply]
We currently only support retries for requests that have no payloads; and gRPC requests necessarily have payloads.

I think we could reasonably support retrying requests that fit into some small buffer capacity... Ideally no more than a few KB.

#4366 would also cover many `UNAVAILABLE` cases automatically.

**Reply by vaibhavsawhney** [reply]
Thank you both for the response. I guess this discussion can be concluded by saying that we cannot do cherry-picking of status codes in gRPC to retry the linked requests currently.

Let me know if I have understood this correctly and I will mark this as the answer and close this discussion.
